src/core/namespace.c

```diff
@@ -287,12 +288,19 @@ static int apply_mount(
 
                 /* First, get rid of everything that is below if there
                  * is anything... Then, overmount it with an
-                 * inaccessible directory. */
+                 * inaccessible path. */
                 umount_recursive(m->path, 0);
 
-                what = "/run/systemd/inaccessible";
-                break;
+                r = lstat(m->path, &target);
+                if (r != 0)
+                        return -errno;
 
+                what = mode_to_inaccessible_node(target.st_mode);
+                if (what == NULL) {
+                        log_error("File type not supported. Note that symlinks are not allowed");
+                        return -1;
```

Comment:
please do not make up error numbers. We try to strictly return negative errnos on error, and your own patch does this correctly a few lines up. You should do the same here, in particular as "return -1" is otherwise misundestood as "return -EPERM" (as EPERM is defined to 1), and that is confusing since this really isn't a permission problem here. Please return -ELOOP or so (after all, that's usually what is returned if something is a symlink but shouldn't be, see O_NOFOLLOW and so on).